### PR TITLE
feat: add Constraint helpers (e.g. from_lengths)

### PR DESCRIPTION
**Pull request by joshka**
Adds helper methods that convert from iterators of u16 values to the
specific Constraint type. This makes it easy to create constraints like:

```rust
// a fixed layout
let constraints = Constraint::from_lengths([10, 20, 10]);

// a centered layout
let constraints = Constraint::from_ratios([(1, 4), (1, 2), (1, 4)]);
let constraints = Constraint::from_percentages([25, 50, 25]);

// a centered layout with a minimum size
let constraints = Constraint::from_mins([0, 100, 0]);

// a sidebar / main layout with maximum sizes
let constraints = Constraint::from_maxes([30, 200]);
```

### CODE REVIEW

**Review by Valentin271 — APPROVED** [review]

**Comment by joshka** [comment]
Needs to be rebased before merging

**Comment by TieWay59** [comment]
Except for the point I mentioned above, I have successfully rebased this branch. I may not solve every conflict as expected, so I prefer not to share my version here. 

@joshka  but I believe most of them can transfer.

**Comment by joshka** [comment]
I'm going to split this into a second PR - the first just a pure refactor (because the underlying changes in layout.rs make it difficult to rebase it and ensure that no code is lost).

File: src/layout.rs [hunk 1]
```diff
@@ -103,7 +103,34 @@ pub enum Alignment {
...
+///
+/// Constraints are used to define the size of a layout. They can be used to define a fixed size, a
+/// percentage of the available space, a ratio of the available space, or a minimum or maximum size.
+///
+/// Relative constraints (percentage, ratio) are calculated relative to the entire space being
+/// split, not the space available after applying the more fixed constraints (min, max, length).
+///
+/// # Examples
+///
+/// `Constraint`` has some helper methods to create lists of constraints from anything that can be
```
Comment (Valentin271) [inline src/layout.rs:116]:
```suggestion
/// `Constraint` has some helper methods to create lists of constraints from anything that can be
```

**Review by Valentin271 — APPROVED** [review]
Other than the typo looks good

**Review by orhun — APPROVED** [review]

**Pull request merged**